File: Find_Peak_Element.py
# Time Complexity = O(logn)
# Space Complexity = O(1)
# Leetcode = yes
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        l, r = 0, len(nums)-1
        while l <= r:
            mid = l + (( r-l) // 2)
            # as we have greater element on mid - 1 position we can find peak at left side
            if nums[mid-1] > nums[mid] and mid > 0 :
                r = mid - 1
            # as mid+1 > mid we can find peak at right side of mid 
            elif mid < len(nums)-1 and nums[mid] < nums[mid +1]:
                l = mid + 1
            else:
                return mid

File: Find Peak Element.cpp
// Time Complexity : O(logn)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no

class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        if(nums.size() == 0) return -1;
        int low = 0;
        int high = nums.size() - 1;
        while(low <= high) {
            int mid = (high - low) / 2 + low;
            // check for peak and also the edge cases
            if((mid == 0 || nums[mid] > nums[mid - 1]) && (mid == nums.size() - 1 || nums[mid] > nums[mid + 1])) 
                return mid;
            // else move towards the bigger element, since nums[i] != nums[i + 1] & nums[-1] = nums[n] = -∞, we will find a peak
            // take care of mid else runtime error
            else if(mid > 0 && nums[mid] < nums[mid - 1]) high = mid - 1;
            else low = mid + 1;
        }
        return -1;
    }
};